### PR TITLE
Issues/issue773

### LINKED ISSUE #773: Simple Plugin Example Problems
I'm following the documentation found here to make a simple plugin.

https://volatility3.readthedocs.io/en/latest/simple-plugin.html

Before I write my own, I'm just going through the steps in the documentation to make the example. However, I have run into two problems. First, the imports shown in the example are incomplete. The resulting plugin will not run without the following:

```python
from volatility3.framework.configuration import requirements
from volatility3.plugins.windows import pslist
```

Once those are added, the plugin runs, but then halts with the following error:

```
# vol -f test.raw windows.testplugin
Volatility 3 Framework 2.0.1
Progress:  100.00		PDB scanning finished                     
Unsatisfied requirement plugins.DllList.pslist: 
Unable to validate the plugin requirements: ['plugins.DllList.pslist']
```

The code that I'm using to test with is cut and pasted together from each step in the example documentation as follows. There are no deviations from the documentation except for the two additional imports identified above.

```python
from volatility3.framework import interfaces
from volatility3.framework.configuration import requirements
from volatility3.plugins.windows import pslist

class DllList(interfaces.plugins.PluginInterface):
    @classmethod
    def get_requirements(cls):
        return [requirements.TranslationLayerRequirement(name = 'primary',
                                                         description = 'Memory layer for the kernel',
                                                         architectures = ["Intel32", "Intel64"]),
                requirements.SymbolTableRequirement(name = "nt_symbols",
                                                    description = "Windows kernel symbols"),
                requirements.PluginRequirement(name = 'pslist',
                                               plugin = pslist.PsList,
                                               version = (1, 0, 0)),
                requirements.ListRequirement(name = 'pid',
                                             element_type = int,
                                             description = "Process IDs to include (all other processes are excluded)",
                                             optional = True)]

    def run(self):
        filter_func = pslist.PsList.create_pid_filter(self.config.get('pid', None))

        return renderers.TreeGrid([("PID", int),
                                   ("Process", str),
                                   ("Base", format_hints.Hex),
                                   ("Size", format_hints.Hex),
                                   ("Name", str),
                                   ("Path", str)],
                                  self._generator(pslist.PsList.list_processes(self.context,
                                                                               self.config['primary'],
                                                                               self.config['nt_symbols'],
                                                                               filter_func = filter_func)))

    def _generator(self, procs):
        for proc in procs:
            for entry in proc.load_order_modules():
                BaseDllName = FullDllName = renderers.UnreadableValue()
                try:
                    BaseDllName = entry.BaseDllName.get_string()
                    # We assume that if the BaseDllName points to an invalid buffer, so will FullDllName
                    FullDllName = entry.FullDllName.get_string()
                except exceptions.InvalidAddressException:
                    pass

                yield (0, (proc.UniqueProcessId,
                           proc.ImageFileName.cast("string", max_length = proc.ImageFileName.vol.count,
                                                   errors = 'replace'),
                           format_hints.Hex(entry.DllBase), format_hints.Hex(entry.SizeOfImage),
                           BaseDllName, FullDllName))
```

What is missing from the example to get it to run?

### PR DESCRIPTION
**Pull request by ikelos**
This updates the simple plugin documentation to use the latest version of the framework, and features available with that (such as the `ModuleRequirement`).

Resolves #773 
Obsoletes #774

### CODE REVIEW

**Comment by ikelos** [comment]
@digitalisx thanks for #774 , hopefully this is a slightly more filled out fix for the documentation issue.  Hopefully this all makes sense, but I'd very much appreciate your eagle eye to see if I've introduced any mistakes or gotten anything wrong...  5;)

File: doc/source/simple-plugin.rst [hunk 1]
```diff
@@ -30,89 +30,132 @@ to be able to run properly.  Any that are defined as optional need not necessari
...
+            return [requirements.ModuleRequirement(name = 'kernel', description = 'Windows kernel',
+                                                   architectures = ["Intel32", "Intel64"]),
                     requirements.ListRequirement(name = 'pid',
                                                  element_type = int,
                                                  description = "Process IDs to include (all other processes are excluded)",
-                                                 optional = True)]
+                                                 optional = True),
+                    requirements.PluginRequirement(name = 'pslist',
+                                                   plugin = pslist.PsList,
+                                                   version = (1, 0, 0))]
```
Comment (digitalisx) [inline doc/source/simple-plugin.rst]: I think the minimum version of the `pslist` plugin should be `2.0.0`. Setting this to `1.0.0` will result in the error mentioned in the issue (#773).

File: doc/source/simple-plugin.rst [hunk 2]
```diff
@@ -30,89 +30,132 @@ to be able to run properly.  Any that are defined as optional need not necessari
...
+The next requirement is a List Requirement, populated by integers.  The description will be presented to the user to
+describe what the value represents.  The optional flag indicates that the plugin can function without the ``pid`` value
+being defined within the configuration tree at all.
 
 ::
 
     requirements.PluginRequirement(name = 'pslist',
                                    plugin = pslist.PsList,
-                                   version = (1, 0, 0)),
+                                   version = (1, 0, 0))
```
Comment (digitalisx) [inline doc/source/simple-plugin.rst]: Same point of above review comment.

**Review by digitalisx — COMMENTED** [review]
@ikelos No problem!
I'm so happy if both the community and the user can go in a better direction.
Thank you for working on the issue on your busy schedule!

I left some comments through the review, could you take a look? :)

File: doc/source/simple-plugin.rst (hunk 1 above)
Comment (digitalisx) [reply doc/source/simple-plugin.rst]: If unintended, this would be a new issue.

**Comment by digitalisx** [comment]
Duplicate of #774

File: doc/source/simple-plugin.rst (hunk 2 above)
Comment (digitalisx) [reply doc/source/simple-plugin.rst]:
```suggestion
                                   version = (2, 0, 0))
```

File: doc/source/simple-plugin.rst (hunk 1 above)
Comment (digitalisx) [reply doc/source/simple-plugin.rst]:
```suggestion
                                                   version = (2, 0, 0))]
```

File: doc/source/simple-plugin.rst (hunk 2 above)
Comment (ikelos) [reply doc/source/simple-plugin.rst]: Thanks, I fixed this manually.  The comma was intentional (to show it was copied from above), but actually it should have been a `]`.

**Review by digitalisx — APPROVED** [review]
@ikelos Good! Thank you for reflecting my opinion. I hope this document will help more people. :)

**Comment by ikelos** [comment]
Thanks for checking it through, it's much appreciated!  5:D

**Comment by utkonos** [comment]
@ikelos @digitalisx Thank you both!

**Comment by utkonos** [comment]
Followup question: I see that this documentation change was merged into develop. Was this intentional? I am unfamiliar with the documentation release process here, but the published documentation is still the old `simple-plugin.rst`. If this is the correct release process, then disregard and forgive my ignorance.

**Comment by utkonos** [comment]
Also, I'm sorry to be such a pest, but this plugin example still doesn't work. There are three imports that I've identified that need to be added to the example.

```python
from volatility3.framework import renderers, interfaces
from volatility3.framework.configuration import requirements
from volatility3.plugins.windows import pslist
```

However, I'm still hitting an exception. I'm looking now to find this import in other plugins.

```python
Volatility 3 Framework 2.0.1
Traceback (most recent call last):B scanning finished                        
  File "/home/vagrant/venv/bin/vol", line 8, in <module>
    sys.exit(main())
  File "/home/vagrant/venv/lib/python3.10/site-packages/volatility3/cli/__init__.py", line 625, in main
    CommandLine().run()
  File "/home/vagrant/venv/lib/python3.10/site-packages/volatility3/cli/__init__.py", line 333, in run
    renderers[args.renderer]().render(constructed.run())
  File "/home/vagrant/venv/lib/python3.10/site-packages/volatility3/plugins/windows/test.py", line 28, in run
    ("Base", format_hints.Hex),
NameError: name 'format_hints' is not defined
```

When someone has read the documentation for the simple plugin, they should not need to find imports to get the example to work. Everything should work when the page has been read and followed.

**Comment by digitalisx** [comment]
Hello @utkonos, Thank you for your continuous tracking of the issue.

Some of the processes and intentions of the document release can be found here. (#767)

And `format_hints` can import this code,


```python
from volatility3.framework.renderers import format_hints
```

That's right, I need to mention that part, but I'm worried about how to appear according to the order of description in the document. If there is no work of the maintainer, I will submit the PR in the same way as last time.

**Comment by ikelos** [comment]
Hi @utkonos, the documentation should match the version that it's released with, and develop tracks with [latest](https://volatility3.readthedocs.io/en/latest/simple-plugin.html).  By default, we go to the latest [stable release](https://volatility3.readthedocs.io/en/stable/simple-plugin.html) whose code doesn't contain the changes and would therefore confuse people if the documentation changed.

The example code wasn't designed to be copied verbatim, and in fact should closely match the existing [`dlllist` plugin](https://github.com/volatilityfoundation/volatility3/blob/develop/volatility3/framework/plugins/windows/dlllist.py) if a copy of the code is needed to work along with.  For complete and functioning plugins, the plugins directory https://github.com/volatilityfoundation/volatility3/tree/develop/volatility3/framework/plugins contains all the examples a user could want.  This example is special because it talks through what each part of the code does.  It was supposed to demonstrate and talk through the key components necessary for writing a plugin, and assumes knowledge of python to be able to fill in parts like the imports, since basic python knowledge is seen as a prerequisite for writing plugins.

Having said all that, I'm happy for the imports to be added as long as there aren't too many that then detract/distract from the key learning points of the example, but the documentation is a teaching resource first and foremost, and copy/pastable code second.

**Comment by utkonos** [comment]
@ikelos That release process is understandable. However, the original issue that I submitted was due to the documentation _not_ being updated to match the released version. The issue that I opened was focused on the published documentation not working for the version that was released (imports totally aside).

**Comment by utkonos** [comment]
One more point: this simple plugin documentation is the very first document that discusses writing a plugin in the documentation. If you want people to know that all the code needed to understand plugins is in that directory, this fact should be somewhere in this documentation. Maybe at the top mention that:

```
The example code wasn't designed to be copied verbatim, and in fact should closely match the existing [dlllist plugin](https://github.com/volatilityfoundation/volatility3/blob/develop/volatility3/framework/plugins/windows/dlllist.py) if a copy of the code is needed to work along with. For complete and functioning plugins, the plugins directory https://github.com/volatilityfoundation/volatility3/tree/develop/volatility3/framework/plugins contains all the examples a user could want.
```

That sentence and the link to the location for a new plugin writer to read the code of existing plugins will fill the gap well enough if the example is not meant to result in a working plugin verbatim.

**Comment by ikelos** [comment]
I understand, unfortunately we can't easily backport fixes for documentation, since the release contains the documentation itself, so to change the documentation we'd need to issue a new release.  I don't think the change warrants a emergency release, but the good news is that we are aiming to roll a new release in September which will therefore come with the fixed up documentation.

Point taken about the documentation, I'll get that updated to reflect that the code isn't complete.

**Comment by utkonos** [comment]
@ikelos Thanks! And sorry if I'm being a pest!

**Comment by ikelos** [comment]
Not at all, thanks for helping us make the project better!  5:D  Hopefully it's clearer from commit `951a0f5d` where to go for samples of the code.  5:)

**Comment by utkonos** [comment]
I've run through the example and my plugin now works nicely. Thanks for all of this. I'm now going to start making some new plugins.

On a side note, I rewrote the plugin a bit more in my style of code. None of which is important for this example. There is one small modification that I did that makes the code more Pythonic. There is a `try` / `except` that can be smoother using `contextlib`. Here is the former code:

```python
try:
    BaseDllName = entry.BaseDllName.get_string()
    # We assume that if the BaseDllName points to an invalid buffer, so will FullDllName
    FullDllName = entry.FullDllName.get_string()
except exceptions.InvalidAddressException:
    pass
```

This is how I rewrote it:

```python
with contextlib.suppress(exceptions.InvalidAddressException):
    BaseDllName = entry.BaseDllName.get_string()
    FullDllName = entry.FullDllName.get_string()
```

The comment can be left in as well, if that's important.

```python
# We assume that if the BaseDllName points to an invalid buffer, so will FullDllName
with contextlib.suppress(exceptions.InvalidAddressException):
    BaseDllName = entry.BaseDllName.get_string()
    FullDllName = entry.FullDllName.get_string()
```

**Comment by ikelos** [comment]
Nice, I'll have to try and remember that one.  I'm a little cautious about making it easier for people to skip past exceptions, it's been somewhat of an uphill battle to get people to properly catch and deal with exceptions, but I guess searching for `contextlib.suppress` is as easy as searching for `pass`.  I'll still try and keep a limit on their use because generally when something goes wrong we should be logging it so the user knows that something when wrong and what but as here, there are certain situations where the exception's already been handled.  I'll have a think about going through the codebase sometime and adding it to our style, thanks!  5:)

**Pull request merged**